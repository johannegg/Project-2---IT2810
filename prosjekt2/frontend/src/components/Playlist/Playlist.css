.playlist-card {
    border: 2px solid #000;
    border-radius: 12px;
    padding: 20px;
    width: 150px;
    height: 100px;
    text-align: center;
    cursor: pointer;
    transition: border-color 0.3s ease;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    box-shadow: 0 4px 8px rgba(0, 0, 0, 0.2);
  }
  
 .playlist-icon{
    font-size: 40px;
 }

 @media (prefers-color-scheme: dark) {
   .playlist-card{
      box-shadow: 0 4px 8px rgba(129, 125, 125, 0.931);
   }
 }
  
  