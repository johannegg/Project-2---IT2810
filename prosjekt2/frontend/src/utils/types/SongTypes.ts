export type Artist = {
    id: string;
    name: string;
};

export type Genre = {
    name: string;
};

export type SongData = {
    id: string;
    title: string;
    views: number;
    year: number;
    artist: Artist;
    genre: Genre;
    lyrics: string;
};