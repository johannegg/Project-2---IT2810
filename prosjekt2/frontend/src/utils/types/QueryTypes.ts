import { SongData } from "./SongTypes";

// Define types for the GraphQL response
export type SongsQueryResponse = {
    songs: SongData[];
};