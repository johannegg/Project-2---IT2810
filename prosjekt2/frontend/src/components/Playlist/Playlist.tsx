import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import "./Playlist.css";
import { SongData } from "../../utils/types/SongTypes";

interface PlaylistProps {
	id: string;
	name: string;
	backgroundColor: string;
	icon: string;
	songs: SongData[];
	onClick: () => void;
	tabIndex?: number;
	onKeyDown?: (event: React.KeyboardEvent) => void;
}

// Color mapping moved outside component for better performance
const colorMapping: Record<string, string> = {
	"#ffffff": "#8a8587",
	"#e8dff5": "#866f95",
	"#fce1e4": "#9e3369",
	"#fcf4dd": "#d7ba28",
	"#ddedea": "#35693f",
	"#daeaf6": "#445988",
	"#8a8587": "#ffffff",
	"#866f95": "#e8dff5",
	"#9e3369": "#fce1e4",
	"#d7ba28": "#fcf4dd",
	"#35693f": "#ddedea",
	"#445988": "#daeaf6",
};

const Playlist: React.FC<PlaylistProps> = ({
	id,
	name,
	backgroundColor,
	icon,
	songs,
	tabIndex,
}) => {
	const navigate = useNavigate();
	const [currentBackgroundColor, setCurrentBackgroundColor] = useState(backgroundColor);

	useEffect(() => {
		const updateBackgroundColor = () => {
			const isDarkModeActive = window.matchMedia("(prefers-color-scheme: dark)").matches;
			const targetColor =
				isDarkModeActive && backgroundColor in colorMapping
					? colorMapping[backgroundColor]
					: backgroundColor;

			setCurrentBackgroundColor(targetColor);
		};

		updateBackgroundColor();
		const colorSchemeMedia = window.matchMedia("(prefers-color-scheme: dark)");
		colorSchemeMedia.addEventListener("change", updateBackgroundColor);

		return () => colorSchemeMedia.removeEventListener("change", updateBackgroundColor);
	}, [backgroundColor]);

	const handleClick = () => {
		navigate(`/playlist/${id}`, {
			state: { playlist: { id, name, backgroundColor: currentBackgroundColor, icon, songs } },
		});
	};

	return (
		<article
			className="playlist-card"
			onClick={handleClick}
			tabIndex={tabIndex}
			onKeyDown={(e) => {
				if (e.key === "Enter" || e.key === " ") handleClick();
			}}
			style={{ backgroundColor: currentBackgroundColor }}
			role="button"
		>
			<div className="playlist-icon">{icon}</div>
			<h3>{name}</h3>
		</article>
	);
};

export default Playlist;
