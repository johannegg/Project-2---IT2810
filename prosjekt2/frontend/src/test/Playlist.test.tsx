import React from "react";
import { describe, test, vi, beforeAll, expect } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Playlist from "../components/Playlist/Playlist";
import { SongData } from "../utils/types/SongTypes";
import "@testing-library/jest-dom";

const mockNavigate = vi.fn();

vi.mock("react-router-dom", async () => {
	const actual = await vi.importActual("react-router-dom");
	return {
		...actual,
		useNavigate: () => mockNavigate,
	};
});

beforeAll(() => {
	const listeners: Array<(ev: MediaQueryListEvent) => unknown> = [];

	vi.spyOn(window, "matchMedia").mockImplementation((query) => ({
		matches: query === "(prefers-color-scheme: dark)",
		media: query,
		onchange: null,
		addEventListener: (event: string, listener: EventListenerOrEventListenerObject) => {
			if (event === "change" && typeof listener === "function") {
				listeners.push(listener);
			}
		},
		removeEventListener: (event: string, listener: EventListenerOrEventListenerObject) => {
			if (event === "change" && typeof listener === "function") {
				const index = listeners.indexOf(listener);
				if (index > -1) listeners.splice(index, 1);
			}
		},
		dispatchEvent: (event: Event): boolean => {
			listeners.forEach((listener) => listener(event as MediaQueryListEvent));
			return true;
		},
		addListener: (callback: (this: MediaQueryList, ev: MediaQueryListEvent) => unknown) => {
			listeners.push(callback);
		},
		removeListener: (callback: (this: MediaQueryList, ev: MediaQueryListEvent) => unknown) => {
			const index = listeners.indexOf(callback);
			if (index > -1) listeners.splice(index, 1);
		},
	}));
});

describe("Playlist Component", () => {
	const mockSongs: SongData[] = [
		{
			id: "1",
			title: "Song 1",
			artist: { id: "1", name: "Artist 1" },
			views: 1000,
			year: 2023,
			genre: { name: "Test Genre 1" },
			lyrics: "La la la la...",
		},
		{
			id: "2",
			title: "Song 2",
			artist: { id: "2", name: "Artist 2" },
			views: 2000,
			year: 2024,
			genre: { name: "Test Genre 2" },
			lyrics: "La la la...",
		},
	];

	const defaultProps = {
		id: "playlist1",
		name: "My Playlist",
		backgroundColor: "#ffffff",
		icon: "ðŸŽµ",
		songs: mockSongs,
		tabIndex: 0,
		onClick: () => {},
	};

	test("renders name and icon", () => {
		const { asFragment } = render(
			<MemoryRouter>
				<Playlist {...defaultProps} />
			</MemoryRouter>,
		);

		expect(screen.getByText("My Playlist")).toBeInTheDocument();
		expect(screen.getByText("ðŸŽµ")).toBeInTheDocument();

		expect(asFragment()).toMatchSnapshot();
	});

	test("navigates with correct state including updated background color", () => {
		const mockSetCurrentBackgroundColor = vi.fn();

		vi.spyOn(React, "useState").mockImplementationOnce(() => [
			"#8a8587", 
			mockSetCurrentBackgroundColor,
		]);

		const { asFragment } = render(
			<MemoryRouter>
				<Playlist {...defaultProps} />
			</MemoryRouter>,
		);

		const playlistCard = screen.getByRole("button");
		fireEvent.click(playlistCard);

		expect(mockNavigate).toHaveBeenCalledWith("/playlist/playlist1", {
			state: {
				playlist: {
					id: "playlist1",
					name: "My Playlist",
					backgroundColor: "#8a8587",
					icon: "ðŸŽµ",
					songs: mockSongs,
				},
			},
		});

		expect(asFragment()).toMatchSnapshot();
	});

	test("triggers navigation with dynamically updated background color", () => {
		const { asFragment, rerender } = render(
			<MemoryRouter>
				<Playlist {...defaultProps} />
			</MemoryRouter>,
		);

		vi.spyOn(window.matchMedia("(prefers-color-scheme: dark)"), "matches", "get").mockReturnValue(
			true,
		);

		rerender(
			<MemoryRouter>
				<Playlist {...defaultProps} />
			</MemoryRouter>,
		);

		const playlistCard = screen.getByRole("button");
		fireEvent.click(playlistCard);

		expect(mockNavigate).toHaveBeenCalledWith("/playlist/playlist1", {
			state: {
				playlist: {
					id: "playlist1",
					name: "My Playlist",
					backgroundColor: "#8a8587",
					icon: "ðŸŽµ",
					songs: mockSongs,
				},
			},
		});

		expect(asFragment()).toMatchSnapshot();
	});

	test("renders the correct number of songs in the playlist", () => {
		const { asFragment } = render(
			<MemoryRouter>
				<Playlist {...defaultProps} />
			</MemoryRouter>,
		);

		const playlistCard = screen.getByRole("button");
		expect(playlistCard).toHaveTextContent("My Playlist");
		expect(mockSongs.length).toBe(2);

		expect(asFragment()).toMatchSnapshot();
	});

	test("handles tabIndex correctly", () => {
		const { asFragment } = render(
			<MemoryRouter>
				<Playlist {...defaultProps} tabIndex={0} />
			</MemoryRouter>,
		);

		const playlistCard = screen.getByRole("button");
		expect(playlistCard).toHaveAttribute("tabIndex", "0");

		expect(asFragment()).toMatchSnapshot();
	});
});
