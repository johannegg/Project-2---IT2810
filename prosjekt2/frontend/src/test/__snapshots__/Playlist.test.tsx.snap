// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Playlist Component > handles tabIndex correctly 1`] = `
<DocumentFragment>
  <article
    class="playlist-card"
    role="button"
    style="background-color: rgb(138, 133, 135);"
    tabindex="0"
  >
    <div
      aria-hidden="true"
      class="playlist-icon"
    >
      🎵
    </div>
    <h3
      aria-label="Playlist name: My Playlist"
    >
      My Playlist
    </h3>
  </article>
</DocumentFragment>
`;

exports[`Playlist Component > navigates with correct state including updated background color 1`] = `
<DocumentFragment>
  <article
    class="playlist-card"
    role="button"
    style="background-color: rgb(138, 133, 135);"
    tabindex="0"
  >
    <div
      aria-hidden="true"
      class="playlist-icon"
    >
      🎵
    </div>
    <h3
      aria-label="Playlist name: My Playlist"
    >
      My Playlist
    </h3>
  </article>
</DocumentFragment>
`;

exports[`Playlist Component > renders name and icon 1`] = `
<DocumentFragment>
  <article
    class="playlist-card"
    role="button"
    style="background-color: rgb(138, 133, 135);"
    tabindex="0"
  >
    <div
      aria-hidden="true"
      class="playlist-icon"
    >
      🎵
    </div>
    <h3
      aria-label="Playlist name: My Playlist"
    >
      My Playlist
    </h3>
  </article>
</DocumentFragment>
`;

exports[`Playlist Component > renders the correct number of songs in the playlist 1`] = `
<DocumentFragment>
  <article
    class="playlist-card"
    role="button"
    style="background-color: rgb(138, 133, 135);"
    tabindex="0"
  >
    <div
      aria-hidden="true"
      class="playlist-icon"
    >
      🎵
    </div>
    <h3
      aria-label="Playlist name: My Playlist"
    >
      My Playlist
    </h3>
  </article>
</DocumentFragment>
`;

exports[`Playlist Component > triggers navigation with dynamically updated background color 1`] = `
<DocumentFragment>
  <article
    class="playlist-card"
    role="button"
    style="background-color: rgb(138, 133, 135);"
    tabindex="0"
  >
    <div
      aria-hidden="true"
      class="playlist-icon"
    >
      🎵
    </div>
    <h3
      aria-label="Playlist name: My Playlist"
    >
      My Playlist
    </h3>
  </article>
</DocumentFragment>
`;
